File: test/hardware/integration/test_arithmetic.py
import cocotb
from cocotb.triggers import Timer, RisingEdge
from cocotb.clock import Clock
from pathlib import Path
from test.driver import run_hardware_test
from test.env import get_all_rtl_files


# Machine Code for Arithmetic Operations
# 0: ADDI x1, x0, 10  (x1 = 10)
# 4: ADDI x2, x0, 5   (x2 = 5)
# 8: ADD x3, x1, x2   (x3 = 15)
# C: SUB x4, x1, x2   (x4 = 5)
# 10: AND x5, x1, x2  (x5 = 0)
# 14: OR x6, x1, x2   (x6 = 15)
# 18: XOR x7, x1, x2  (x7 = 15)
# 1C: SLL x8, x1, x2  (x8 = 320)
# 20: SRL x9, x1, x2  (x9 = 0)
# 24: SLT x10, x2, x1 (x10 = 1)
# 28: EBREAK          (Stop)
PROGRAM = [
    "00a00093", # ADDI x1, x0, 10
    "00500113", # ADDI x2, x0, 5
    "002081b3", # ADD x3, x1, x2
    "40208233", # SUB x4, x1, x2
    "0020f2b3", # AND x5, x1, x2
    "0020e333", # OR x6, x1, x2
    "0020c3b3", # XOR x7, x1, x2
    "00209433", # SLL x8, x1, x2
    "002054b3", # SRL x9, x1, x2
    "00112533", # SLT x10, x2, x1
    "00100073", # EBREAK
    "00000013", # NOP
    "00000013", # NOP
]

def create_hex_file(filename="program.hex"):
    with open(filename, "w") as f:
        for instr in PROGRAM:
            f.write(f"{instr}\n")

@cocotb.test()
async def test_arithmetic_program(dut):
    """
    Run arithmetic operations test.
    """
    create_hex_file("program.hex")

    clock = Clock(dut.clk, 10, units="ns")
    cocotb.start_soon(clock.start())

    dut.rst_n.value = 0
    await Timer(20, units="ns")
    dut.rst_n.value = 1
    
    # Wait for execution
    for i in range(1000):
        await RisingEdge(dut.clk)
        try:
            pc_ex = dut.u_core.u_backend.id_ex_program_counter.value.integer
        except:
            pc_ex = 0

        if pc_ex == 40: # EBREAK address (0x28)
            dut._log.info("EBREAK Executed. Stopping.")
            for _ in range(100):
                await RisingEdge(dut.clk)
            break
            
    # Verify State
    try:
        x1 = dut.u_core.u_backend.u_regfile.registers[1].value.integer
        x2 = dut.u_core.u_backend.u_regfile.registers[2].value.integer
        x3 = dut.u_core.u_backend.u_regfile.registers[3].value.integer
        x4 = dut.u_core.u_backend.u_regfile.registers[4].value.integer
        x5 = dut.u_core.u_backend.u_regfile.registers[5].value.integer
        x6 = dut.u_core.u_backend.u_regfile.registers[6].value.integer
        x7 = dut.u_core.u_backend.u_regfile.registers[7].value.integer
        x8 = dut.u_core.u_backend.u_regfile.registers[8].value.integer
        x9 = dut.u_core.u_backend.u_regfile.registers[9].value.integer
        x10 = dut.u_core.u_backend.u_regfile.registers[10].value.integer
        
        assert x1 == 10, f"x1 should be 10, got {x1}"
        assert x2 == 5, f"x2 should be 5, got {x2}"
        assert x3 == 15, f"x3 should be 15, got {x3}"
        assert x4 == 5, f"x4 should be 5, got {x4}"
        assert x5 == 0, f"x5 should be 0, got {x5}"
        assert x6 == 15, f"x6 should be 15, got {x6}"
        assert x7 == 15, f"x7 should be 15, got {x7}"
        assert x8 == 320, f"x8 should be 320, got {x8}"
        assert x9 == 0, f"x9 should be 0, got {x9}"
        assert x10 == 1, f"x10 should be 1, got {x10}"
        
    except Exception as e:
        dut._log.error(f"Failed to inspect registers: {e}")
        raise e


def test_arithmetic():
    run_hardware_test(
        module_name=Path(__file__).stem,
        toplevel="chip_top",
        verilog_sources=get_all_rtl_files()
    )

File: test/hardware/integration/test_arithmetic.cpp
// Test: Arithmetic Operations Integration Test
// Runs arithmetic operations on the full chip:
// - ADDI x1, x0, 10  (x1 = 10)
// - ADDI x2, x0, 5   (x2 = 5)
// - ADD x3, x1, x2   (x3 = 15)
// - SUB x4, x1, x2   (x4 = 5)
// - AND x5, x1, x2   (x5 = 0)
// - OR x6, x1, x2    (x6 = 15)
// - XOR x7, x1, x2   (x7 = 15)
// - SLL x8, x1, x2   (x8 = 320)
// - SRL x9, x1, x2   (x9 = 0)
// - SLT x10, x2, x1  (x10 = 1)
// - EBREAK           (Stop)

#include "../common/tb_base.h"
#include <Vchip_top.h>
#include <Vchip_top___024root.h>

class ChipTopTestbench : public ClockedTestbench<Vchip_top> {
public:
    ChipTopTestbench() : ClockedTestbench<Vchip_top>(100, true, "dump.vcd") {
        dut->rst_n = 0;
    }

    void set_clk(uint8_t value) override {
        dut->clk = value;
    }

    void load_program(const std::vector<uint32_t>& program) {
        for (size_t i = 0; i < program.size(); i++) {
            dut->rootp->chip_top__DOT__u_memory_subsystem__DOT__u_main_memory__DOT__memory[i] = program[i];
        }
    }

    uint32_t read_register(int reg_idx) {
        if (reg_idx < 0 || reg_idx >= 32) return 0;
        return dut->rootp->chip_top__DOT__u_tile_0__DOT__u_core__DOT__u_backend__DOT__u_regfile__DOT__registers[reg_idx];
    }

    uint32_t get_pc_ex() {
        return dut->rootp->chip_top__DOT__u_tile_0__DOT__u_core__DOT__u_backend__DOT__id_ex_program_counter;
    }

    void do_reset() {
        dut->rst_n = 0;
        for (int i = 0; i < 20; i++) tick();
        dut->rst_n = 1;
        for (int i = 0; i < 5; i++) tick();
    }
};

int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    
    ChipTopTestbench tb;

    std::vector<uint32_t> program = {
        0x00a00093, // ADDI x1, x0, 10
        0x00500113, // ADDI x2, x0, 5
        0x002081b3, // ADD x3, x1, x2
        0x40208233, // SUB x4, x1, x2
        0x0020f2b3, // AND x5, x1, x2
        0x0020e333, // OR x6, x1, x2
        0x0020c3b3, // XOR x7, x1, x2
        0x00209433, // SLL x8, x1, x2
        0x002054b3, // SRL x9, x1, x2
        0x00112533, // SLT x10, x2, x1
        0x00100073, // EBREAK
        0x00000013, // NOP
        0x00000013, // NOP
    };

    tb.load_program(program);
    tb.do_reset();

    // Run until EBREAK (PC = 0x28 = 40)
    bool ebreak_reached = false;
    for (int cycles = 0; cycles < 1000; cycles++) {
        tb.tick();
        
        uint32_t pc_ex = tb.get_pc_ex();
        if (pc_ex == 40) { // EBREAK instruction address
            ebreak_reached = true;
            for (int i = 0; i < 10; i++) tb.tick();
            break;
        }
    }

    TB_ASSERT_EQ(ebreak_reached, true, "EBREAK should be reached");

    // Verify Register Values
    TB_ASSERT_EQ(tb.read_register(1), 10, "x1 should be 10");
    TB_ASSERT_EQ(tb.read_register(2), 5, "x2 should be 5");
    TB_ASSERT_EQ(tb.read_register(3), 15, "x3 should be 15");
    TB_ASSERT_EQ(tb.read_register(4), 5, "x4 should be 5");
    TB_ASSERT_EQ(tb.read_register(5), 0, "x5 should be 0");
    TB_ASSERT_EQ(tb.read_register(6), 15, "x6 should be 15");
    TB_ASSERT_EQ(tb.read_register(7), 15, "x7 should be 15");
    TB_ASSERT_EQ(tb.read_register(8), 320, "x8 should be 320");
    TB_ASSERT_EQ(tb.read_register(9), 0, "x9 should be 0");
    TB_ASSERT_EQ(tb.read_register(10), 1, "x10 should be 1");

    return 0;
}
